File: pipeline.py
"""
Core pipeline engine (optimized for quality).

Flow:
  1. Stage 0: Research gathering (Anthropic)
  2. Stage 1: 2 parallel drafts (Sonnet + GPT-4o)
  3. Judge: Pick best draft + note strengths from loser
  4. Stages 2-5: Four enhancement stages with critique before each:
     - Stage 2: Deep Enhancement (artistic + academic depth)
     - Stage 3: De-AI & Voice Authenticity (strip LLM patterns)
     - Stage 4: Oral Delivery Optimization (breath, rhythm, flow)
     - Stage 5: Final Polish (line-by-line refinement)
  5. Save opening paragraph for future differentiation

Each enhancement stage receives: topic, research brief, critique feedback, previous output.
"""

import json
import os
import re
from concurrent.futures import ThreadPoolExecutor, as_completed
from pathlib import Path

import anthropic
import openai
from dotenv import load_dotenv

from prompts import (
    CRITIQUE_TEMPLATE,
    DIFFERENTIATION_CONTEXT,
    DRAFT_VARIANTS,
    ENHANCEMENT_STAGES,
    JUDGE_PROMPT,
    LEARNING_ADDONS,
    PERSPECTIVE_LENSES,
    get_combined_lens_prompt,
    get_draft_stage,
    get_research_stage,
)

load_dotenv()


def _init_keys():
    """Load API keys from st.secrets (Streamlit Cloud) into env vars."""
    try:
        import streamlit as st
        for key in ("ANTHROPIC_API_KEY", "OPENAI_API_KEY"):
            if key not in os.environ and key in st.secrets:
                os.environ[key] = st.secrets[key]
    except Exception:
        pass


def _get_anthropic_client():
    _init_keys()
    return anthropic.Anthropic()


def _get_openai_client():
    _init_keys()
    return openai.OpenAI()


DEFAULT_ANTHROPIC_MODEL = "claude-sonnet-4-20250514"
DEFAULT_OPENAI_MODEL = "gpt-4o-2024-11-20"
MAX_TOKENS = 16384

# File to store previous speech openings for differentiation
HISTORY_FILE = Path(__file__).parent / "speech_history.json"


def _load_history() -> list[str]:
    if HISTORY_FILE.exists():
        try:
            data = json.loads(HISTORY_FILE.read_text(encoding="utf-8"))
            return data.get("openings", [])
        except (json.JSONDecodeError, KeyError):
            return []
    return []


def _save_opening(text: str):
    openings = _load_history()
    # Store first 300 chars as the "opening paragraph"
    opening = text[:300].strip()
    openings.append(opening)
    # Keep last 20 openings max
    openings = openings[-20:]
    HISTORY_FILE.write_text(
        json.dumps({"openings": openings}, indent=2), encoding="utf-8"
    )


def _call_llm(
    provider: str,
    system: str,
    user_content: str,
    temperature: float = 0.7,
    model_override: str | None = None,
) -> str:
    """Unified LLM call for both Anthropic and OpenAI."""
    if provider == "openai":
        model = model_override or DEFAULT_OPENAI_MODEL
        response = _get_openai_client().chat.completions.create(
            model=model,
            max_tokens=MAX_TOKENS,
            temperature=temperature,
            messages=[
                {"role": "system", "content": system},
                {"role": "user", "content": user_content},
            ],
        )
        return response.choices[0].message.content

    else:  # anthropic
        model = model_override or DEFAULT_ANTHROPIC_MODEL
        message = _get_anthropic_client().messages.create(
            model=model,
            max_tokens=MAX_TOKENS,
            temperature=temperature,
            system=system,
            messages=[{"role": "user", "content": user_content}],
        )
        return message.content[0].text


def _call_llm_safe(provider: str, system: str, user_content: str, **kwargs) -> str:
    """Wrapper with error handling."""
    try:
        return _call_llm(provider, system, user_content, **kwargs)
    except anthropic.RateLimitError:
        raise RuntimeError("Rate limited by Anthropic. Wait a moment and retry.")
    except anthropic.APIConnectionError:
        raise RuntimeError("Cannot reach Anthropic API. Check your network.")
    except anthropic.AuthenticationError:
        raise RuntimeError("Invalid Anthropic API key. Check your .env file.")
    except anthropic.APIError as e:
        raise RuntimeError(f"Anthropic API error: {e.message}")
    except openai.RateLimitError:
        raise RuntimeError("Rate limited by OpenAI. Wait a moment and retry.")
    except openai.APIConnectionError:
        raise RuntimeError("Cannot reach OpenAI API. Check your network.")
    except openai.AuthenticationError:
        raise RuntimeError("Invalid OpenAI API key. Check your .env file.")
    except openai.APIError as e:
        raise RuntimeError(f"OpenAI API error: {e}")


# ──────────────────────────────────────────────
# Stage 0: Research Gathering
# ──────────────────────────────────────────────
def run_research(topic: str, length: str = "10 min") -> str:
    stage = get_research_stage(length)
    user_content = stage["user_template"].format(topic=topic)
    return _call_llm_safe(
        provider=stage["provider"],
        system=stage["system"],
        user_content=user_content,
        temperature=stage["temperature"],
        model_override=stage["model_override"],
    )


# ──────────────────────────────────────────────
# Stage 1: Parallel Drafts
# ──────────────────────────────────────────────
def run_parallel_drafts(topic: str, research: str, length: str = "10 min") -> list[dict]:
    """Generate drafts in parallel. Returns list of {label, text}."""
    # Add differentiation context if we have history
    openings = _load_history()
    diff_prefix = ""
    if openings:
        diff_prefix = DIFFERENTIATION_CONTEXT.format(
            previous_openings="\n---\n".join(openings[-5:])
        )

    stage = get_draft_stage(length)
    base_user = stage["user_template"].format(topic=topic, research=research)
    user_content = diff_prefix + base_user

    num_drafts = len(DRAFT_VARIANTS)
    results = [None] * num_drafts

    def _generate(idx, variant):
        text = _call_llm_safe(
            provider=variant["provider"],
            system=stage["system"],
            user_content=user_content,
            temperature=variant["temperature"],
            model_override=variant["model_override"],
        )
        return idx, variant["label"], text

    with ThreadPoolExecutor(max_workers=num_drafts) as executor:
        futures = [
            executor.submit(_generate, i, v) for i, v in enumerate(DRAFT_VARIANTS)
        ]
        for future in as_completed(futures):
            idx, label, text = future.result()
            results[idx] = {"label": label, "text": text}

    return results


# ──────────────────────────────────────────────
# Judge: Select Best Draft
# ──────────────────────────────────────────────
def run_judge(topic: str, drafts: list[dict]) -> dict:
    """
    Judge drafts (2 or 3). Returns {
        winner_index: int,
        winner_label: str,
        winner_text: str,
        judgment: str,
        borrow_notes: str
    }
    """
    stage = JUDGE_PROMPT

    # Build user content dynamically based on number of drafts
    if len(drafts) == 2:
        user_content = stage["user_template_2"].format(
            topic=topic,
            draft_a=drafts[0]["text"],
            draft_b=drafts[1]["text"],
        )
        letter_map = {"A": 0, "B": 1}
        pattern = r"WINNER:\s*([AB])"
    else:
        user_content = stage["user_template"].format(
            topic=topic,
            draft_a=drafts[0]["text"],
            draft_b=drafts[1]["text"],
            draft_c=drafts[2]["text"],
        )
        letter_map = {"A": 0, "B": 1, "C": 2}
        pattern = r"WINNER:\s*([ABC])"

    judgment = _call_llm_safe(
        provider=stage["provider"],
        system=stage["system"],
        user_content=user_content,
        temperature=stage["temperature"],
        model_override=stage["model_override"],
    )

    # Parse winner from judgment
    winner_letter = "A"  # default fallback
    match = re.search(pattern, judgment, re.IGNORECASE)
    if match:
        winner_letter = match.group(1).upper()

    winner_index = letter_map.get(winner_letter, 0)

    # Extract borrow notes
    borrow_notes = ""
    borrow_match = re.search(
        r"BORROW FROM LOSER[S]?:\s*\n(.*)", judgment, re.DOTALL | re.IGNORECASE
    )
    if borrow_match:
        borrow_notes = borrow_match.group(1).strip()

    return {
        "winner_index": winner_index,
        "winner_label": drafts[winner_index]["label"],
        "winner_text": drafts[winner_index]["text"],
        "judgment": judgment,
        "borrow_notes": borrow_notes,
    }


# ──────────────────────────────────────────────
# Critique (between enhancement stages)
# ──────────────────────────────────────────────
def run_critique(topic: str, text: str, completed_stage: str, next_stage: str) -> str:
    tmpl = CRITIQUE_TEMPLATE
    user_content = tmpl["user_template"].format(
        topic=topic,
        completed_stage=completed_stage,
        next_stage=next_stage,
        text=text,
    )
    return _call_llm_safe(
        provider=tmpl["provider"],
        system=tmpl["system"],
        user_content=user_content,
        temperature=tmpl["temperature"],
        model_override=tmpl["model_override"],
    )


# ──────────────────────────────────────────────
# Enhancement Stages (2-5)
# ──────────────────────────────────────────────
def run_enhancement_stage(
    stage_index: int,
    topic: str,
    research: str,
    critique: str,
    previous_output: str,
) -> str:
    stage = ENHANCEMENT_STAGES[stage_index]
    user_content = stage["user_template"].format(
        topic=topic,
        research=research,
        critique=critique,
        previous_output=previous_output,
    )
    return _call_llm_safe(
        provider=stage["provider"],
        system=stage["system"],
        user_content=user_content,
        temperature=stage["temperature"],
        model_override=stage.get("model_override"),
    )


# ──────────────────────────────────────────────
# Full Pipeline (generator for UI updates)
# ──────────────────────────────────────────────
def run_full_pipeline(topic: str, length: str = "10 min"):
    """
    Generator yielding status updates as tuples:
        (step_name, step_type, data)

    step_type is one of: "research", "drafts", "judge", "critique", "enhancement", "done"
    data contains the relevant output for that step.

    Args:
        topic: The speech topic
        length: Speech length key ("5 min", "10 min", "15 min", "20 min")
    """

    # Step 1: Research
    yield ("Stage 0: Research Gathering", "research", {"status": "running"})
    research = run_research(topic, length)
    yield ("Stage 0: Research Gathering", "research", {"status": "done", "text": research})

    # Step 2: Parallel drafts
    yield ("Stage 1: Parallel Drafts", "drafts", {"status": "running"})
    drafts = run_parallel_drafts(topic, research, length)
    yield ("Stage 1: Parallel Drafts", "drafts", {"status": "done", "drafts": drafts})

    # Step 3: Judge
    yield ("Judge: Select Best Draft", "judge", {"status": "running"})
    judge_result = run_judge(topic, drafts)
    yield ("Judge: Select Best Draft", "judge", {"status": "done", **judge_result})

    current_text = judge_result["winner_text"]

    # Steps 4-7: Enhancement stages with critiques between them
    for i, stage in enumerate(ENHANCEMENT_STAGES):
        # Critique before this stage (except the first enhancement if judge already provided feedback)
        next_stage_name = stage["name"]
        if i == 0:
            prev_stage_name = "Draft Selection (Judge)"
        else:
            prev_stage_name = ENHANCEMENT_STAGES[i - 1]["name"]

        critique_name = f"Critique: {prev_stage_name} → {next_stage_name}"
        yield (critique_name, "critique", {"status": "running"})
        critique = run_critique(topic, current_text, prev_stage_name, next_stage_name)
        yield (critique_name, "critique", {"status": "done", "text": critique})

        # Enhancement stage
        yield (stage["name"], "enhancement", {"status": "running", "stage_index": i})
        current_text = run_enhancement_stage(i, topic, research, critique, current_text)
        yield (stage["name"], "enhancement", {"status": "done", "stage_index": i, "text": current_text})

    # Save opening for future differentiation
    _save_opening(current_text)

    yield ("Complete", "done", {"final_text": current_text})


# ══════════════════════════════════════════════════════════════════════════════
# LENS ANALYSIS (Sovereign Mind / Reflect Mode)
# ══════════════════════════════════════════════════════════════════════════════

def run_lens_analysis(situation: str, selected_lenses: list[tuple[str, str]]) -> str:
    """
    Generate a text-based lens analysis (Quick Insight mode).

    Args:
        situation: The user's situation/dilemma to analyze
        selected_lenses: List of (category_id, lens_id) tuples

    Returns:
        Structured analysis text
    """
    from lenses import (
        LENS_ANALYSIS_SYSTEM,
        LENS_ANALYSIS_USER_TEMPLATE,
        get_lens_prompts,
    )

    lens_prompts = get_lens_prompts(selected_lenses)
    user_content = LENS_ANALYSIS_USER_TEMPLATE.format(
        situation=situation,
        lens_prompts=lens_prompts,
    )

    return _call_llm_safe(
        provider="anthropic",
        system=LENS_ANALYSIS_SYSTEM,
        user_content=user_content,
        temperature=0.7,
        model_override="claude-sonnet-4-20250514",
    )


def run_lens_audio_script(
    situation: str,
    selected_lenses: list[tuple[str, str]],
    minutes: int = 5,
) -> str:
    """
    Generate an audio script for Deep Reflection mode.

    Args:
        situation: The user's situation/dilemma
        selected_lenses: List of (category_id, lens_id) tuples
        minutes: Target audio length in minutes

    Returns:
        Audio script text ready for TTS
    """
    from lenses import (
        LENS_AUDIO_SYSTEM,
        LENS_AUDIO_USER_TEMPLATE,
        get_lens_prompts,
    )

    # Approximately 150 words per minute for thoughtful narration
    word_count = minutes * 150

    lens_prompts = get_lens_prompts(selected_lenses)
    user_content = LENS_AUDIO_USER_TEMPLATE.format(
        situation=situation,
        lens_prompts=lens_prompts,
        word_count=word_count,
        minutes=minutes,
    )

    return _call_llm_safe(
        provider="anthropic",
        system=LENS_AUDIO_SYSTEM,
        user_content=user_content,
        temperature=0.8,
        model_override="claude-sonnet-4-20250514",
    )


# ══════════════════════════════════════════════════════════════════════════════
# LEARNING ADD-ONS (on-demand after episode generation)
# ══════════════════════════════════════════════════════════════════════════════

def generate_addon(addon_key: str, topic: str, transcript: str) -> str:
    """
    Generate a learning add-on (quiz, journal, takeaways).

    Args:
        addon_key: One of "quiz", "journal", "takeaways"
        topic: The episode topic
        transcript: The full episode transcript

    Returns:
        Generated add-on content
    """
    if addon_key not in LEARNING_ADDONS:
        raise ValueError(f"Unknown add-on: {addon_key}")

    addon = LEARNING_ADDONS[addon_key]
    user_content = addon["user_template"].format(topic=topic, transcript=transcript)

    return _call_llm_safe(
        provider=addon.get("provider", "anthropic"),
        system=addon["system"],
        user_content=user_content,
        temperature=addon["temperature"],
        model_override=addon.get("model_override", "claude-sonnet-4-20250514"),
    )


def generate_perspective(lens_key: str, topic: str, transcript: str) -> str:
    """
    Generate a perspective lens analysis.

    Args:
        lens_key: One of the perspective lens keys (stoic, first_principles, etc.)
        topic: The episode topic
        transcript: The full episode transcript

    Returns:
        Generated perspective analysis
    """
    if lens_key not in PERSPECTIVE_LENSES:
        raise ValueError(f"Unknown perspective lens: {lens_key}")

    lens = PERSPECTIVE_LENSES[lens_key]
    user_content = lens["user_template"].format(topic=topic, transcript=transcript)

    return _call_llm_safe(
        provider=lens.get("provider", "anthropic"),
        system=lens["system"],
        user_content=user_content,
        temperature=lens["temperature"],
        model_override=lens.get("model_override", "claude-sonnet-4-20250514"),
    )


def generate_combined_perspectives(lens_keys: list[str], topic: str, transcript: str) -> str:
    """
    Generate a combined analysis from multiple perspective lenses.

    Args:
        lens_keys: List of perspective lens keys to combine
        topic: The episode topic
        transcript: The full episode transcript

    Returns:
        Generated combined perspective analysis
    """
    combined = get_combined_lens_prompt(lens_keys)
    if not combined:
        raise ValueError(f"Invalid lens combination: {lens_keys}")

    user_content = combined["user_template"].format(topic=topic, transcript=transcript)

    return _call_llm_safe(
        provider=combined.get("provider", "anthropic"),
        system=combined["system"],
        user_content=user_content,
        temperature=combined["temperature"],
        model_override=combined.get("model_override", "claude-sonnet-4-20250514"),
    )

File: mindcast-web/src/lib/ai/pipeline.ts
/**
 * MindCast AI Pipeline — v2
 * Server-side episode generation using Claude, GPT-4, and fast models
 */

import Anthropic from '@anthropic-ai/sdk';
import OpenAI from 'openai';
import { withRetry, withTimeout } from '@/lib/async-utils';
import { isRetryableError } from '@/lib/ai/retry';
import {
  getResearchPrompt,
  getDraftPrompt,
  ENHANCEMENT_STAGES,
  JUDGE_PROMPT,
  SUPPORT_CHECK_PROMPT,
  PRONUNCIATION_PROMPT,
  LEARNING_ADDONS,
  PROMPT_VERSION,
  getCanonRemasterPrompt,
  CANON_QUALITY_GATE,
  type EpisodeLength,
} from './prompts';

// Re-export types for consumers
export type { EpisodeLength };

// Clients (lazy initialization)
let anthropicClient: Anthropic | null = null;
let openaiClient: OpenAI | null = null;

function getAnthropicClient(): Anthropic {
  if (!anthropicClient) {
    anthropicClient = new Anthropic({
      apiKey: process.env.ANTHROPIC_API_KEY,
    });
  }
  return anthropicClient;
}

function getOpenAIClient(): OpenAI {
  if (!openaiClient) {
    openaiClient = new OpenAI({
      apiKey: process.env.OPENAI_API_KEY,
    });
  }
  return openaiClient;
}

const DEFAULT_ANTHROPIC_MODEL = process.env.ANTHROPIC_MODEL || 'claude-sonnet-4-5-20250929';
const DEFAULT_OPENAI_MODEL = process.env.OPENAI_MODEL || 'gpt-4o-2024-11-20';
const FAST_OPENAI_MODEL = process.env.OPENAI_FAST_MODEL || 'gpt-4o-mini';
const MAX_TOKENS = 16384;

// ============================================================================
// Core LLM Call
// ============================================================================

type Provider = 'anthropic' | 'openai';

interface LLMCallOptions {
  provider: Provider;
  system: string;
  user: string;
  temperature?: number;
  model?: string;
  maxTokens?: number;
  promptVersion?: string;
  purpose?: string;
}

async function callLLM({
  provider,
  system,
  user,
  temperature = 0.7,
  model,
  maxTokens,
  promptVersion,
  purpose,
}: LLMCallOptions): Promise<string> {
  const resolvedModel = model || (provider === 'openai' ? DEFAULT_OPENAI_MODEL : DEFAULT_ANTHROPIC_MODEL);
  const resolvedMaxTokens = maxTokens ?? MAX_TOKENS;
  console.info('LLM call', {
    provider,
    model: resolvedModel,
    temperature,
    promptVersion,
    purpose,
  });

  if (provider === 'openai') {
    const client = getOpenAIClient();
    const response = await withRetry(
      () =>
        withTimeout(
          client.chat.completions.create({
            model: resolvedModel,
            max_tokens: resolvedMaxTokens,
            temperature,
            messages: [
              { role: 'system', content: system },
              { role: 'user', content: user },
            ],
          }),
          45000,
          'openai-chat'
        ),
      { retries: 2, shouldRetry: isRetryableError, label: 'openai-chat' }
    );
    return response.choices[0]?.message?.content || '';
  }

  // Anthropic
  const client = getAnthropicClient();
  const message = await withRetry(
    () =>
      withTimeout(
        client.messages.create({
          model: resolvedModel,
          max_tokens: resolvedMaxTokens,
          temperature,
          system,
          messages: [{ role: 'user', content: user }],
        }),
        45000,
        'anthropic-chat'
      ),
    { retries: 2, shouldRetry: isRetryableError, label: 'anthropic-chat' }
  );

  const content = message.content[0];
  if (content.type === 'text') {
    return content.text;
  }
  return '';
}

// ============================================================================
// Pipeline Stages
// ============================================================================

export async function runResearch(topic: string, length: EpisodeLength): Promise<string> {
  const prompt = getResearchPrompt(topic, length);
  return callLLM({
    provider: 'anthropic',
    system: prompt.system,
    user: prompt.user,
    temperature: prompt.temperature,
    promptVersion: PROMPT_VERSION,
    purpose: 'research',
  });
}

export async function runDraft(
  topic: string,
  research: string,
  length: EpisodeLength,
  provider: Provider = 'anthropic',
  style?: string
): Promise<string> {
  const prompt = getDraftPrompt(topic, research, length, style);
  return callLLM({
    provider,
    system: prompt.system,
    user: prompt.user,
    temperature: prompt.temperature,
    promptVersion: PROMPT_VERSION,
    purpose: `draft:${provider}`,
  });
}

export async function runParallelDrafts(
  topic: string,
  research: string,
  length: EpisodeLength,
  style?: string
): Promise<{ label: string; text: string }[]> {
  const [anthropicDraft, openaiDraft] = await Promise.all([
    runDraft(topic, research, length, 'anthropic', style),
    runDraft(topic, research, length, 'openai', style),
  ]);

  return [
    { label: 'Claude', text: anthropicDraft },
    { label: 'GPT-4', text: openaiDraft },
  ];
}

export async function runJudge(
  topic: string,
  drafts: { label: string; text: string }[]
): Promise<{
  winnerIndex: number;
  winnerLabel: string;
  winnerText: string;
  judgment: string;
  graftParagraph: string;
}> {
  // Use faster GPT-4o-mini for judging (doesn't need full model power)
  const judgment = await callLLM({
    provider: 'openai',
    model: FAST_OPENAI_MODEL,
    system: JUDGE_PROMPT.system,
    user: JUDGE_PROMPT.userTemplate
      .replace('{topic}', topic)
      .replace('{draftA}', drafts[0].text)
      .replace('{draftB}', drafts[1].text),
    temperature: JUDGE_PROMPT.temperature,
    promptVersion: PROMPT_VERSION,
    purpose: 'judge',
  });

  // Parse winner
  const winnerMatch = judgment.match(/WINNER:\s*([AB])/i);
  const winnerLetter = winnerMatch?.[1]?.toUpperCase() || 'A';
  const winnerIndex = winnerLetter === 'A' ? 0 : 1;

  // Parse graft paragraph (v2 format)
  const graftMatch = judgment.match(/GRAFT_FROM_LOSER:\s*"?([\s\S]*?)"?\s*(?:WHY_WINNER_WINS:|$)/i);
  const graftParagraph = graftMatch?.[1]?.trim() || '';

  return {
    winnerIndex,
    winnerLabel: drafts[winnerIndex].label,
    winnerText: drafts[winnerIndex].text,
    judgment,
    graftParagraph,
  };
}

// ============================================================================
// Support Check (drift prevention)
// ============================================================================

export async function runSupportCheck(
  research: string,
  script: string
): Promise<{ ok: boolean; issues: string }> {
  const result = await callLLM({
    provider: 'openai',
    model: FAST_OPENAI_MODEL,
    system: SUPPORT_CHECK_PROMPT.system,
    user: SUPPORT_CHECK_PROMPT.userTemplate
      .replace('{research}', research)
      .replace('{script}', script),
    temperature: SUPPORT_CHECK_PROMPT.temperature,
    maxTokens: 2048,
    promptVersion: PROMPT_VERSION,
    purpose: 'support-check',
  });

  const ok = result.trim().toUpperCase() === 'OK';
  return { ok, issues: ok ? '' : result };
}

// ============================================================================
// Pronunciation Extraction (TTS metadata)
// ============================================================================

export async function runPronunciation(
  script: string
): Promise<string> {
  return callLLM({
    provider: 'openai',
    model: FAST_OPENAI_MODEL,
    system: PRONUNCIATION_PROMPT.system,
    user: PRONUNCIATION_PROMPT.userTemplate.replace('{script}', script),
    temperature: PRONUNCIATION_PROMPT.temperature,
    maxTokens: 1024,
    promptVersion: PROMPT_VERSION,
    purpose: 'pronunciation',
  });
}

// Enhancement stages run directly without a separate critique step
export async function runEnhancementStage(
  stageIndex: number,
  topic: string,
  _research: string,
  previousOutput: string
): Promise<string> {
  const stage = ENHANCEMENT_STAGES[stageIndex];
  return callLLM({
    provider: 'anthropic',
    system: stage.system,
    user: stage.userTemplate
      .replace('{topic}', topic)
      .replace('{previousOutput}', previousOutput),
    temperature: stage.temperature,
    promptVersion: PROMPT_VERSION,
    purpose: `enhance:${stage.name}`,
  });
}

// ============================================================================
// Full Pipeline
// ============================================================================

export type PipelineStep =
  | { type: 'research'; status: 'running' | 'done'; research?: string }
  | { type: 'drafts'; status: 'running' | 'done'; drafts?: { label: string; text: string }[]; draftA?: string; draftB?: string }
  | { type: 'judge'; status: 'running' | 'done'; winner?: string; winnerText?: string; judgment?: string }
  | { type: 'support-check'; status: 'running' | 'done'; ok?: boolean; issues?: string }
  | { type: 'enhancement'; status: 'running' | 'done'; stageName: string; stageIndex: number; text?: string; enhancedText?: string }
  | { type: 'pronunciation'; status: 'running' | 'done'; pronunciations?: string }
  | { type: 'done'; finalText: string; pronunciations?: string };

export async function* runFullPipeline(
  topic: string,
  length: EpisodeLength = '10 min',
  style?: string
): AsyncGenerator<PipelineStep> {
  // Stage 0: Research
  yield { type: 'research', status: 'running' };
  const research = await runResearch(topic, length);
  yield { type: 'research', status: 'done', research };

  // Stage 1: Parallel Drafts (with style applied)
  yield { type: 'drafts', status: 'running' };
  const drafts = await runParallelDrafts(topic, research, length, style);
  yield { type: 'drafts', status: 'done', drafts, draftA: drafts[0].text, draftB: drafts[1].text };

  // Judge
  yield { type: 'judge', status: 'running' };
  const judgeResult = await runJudge(topic, drafts);
  yield {
    type: 'judge',
    status: 'done',
    winner: judgeResult.winnerLabel,
    winnerText: judgeResult.winnerText,
    judgment: judgeResult.judgment,
  };

  // Support Check (drift prevention — fast model)
  yield { type: 'support-check', status: 'running' };
  const supportResult = await runSupportCheck(research, judgeResult.winnerText);
  yield { type: 'support-check', status: 'done', ok: supportResult.ok, issues: supportResult.issues };

  let currentText = judgeResult.winnerText;

  // OPTIMIZED: Only 2 enhancement stages (down from 4), no separate critique step
  for (let i = 0; i < ENHANCEMENT_STAGES.length; i++) {
    const stage = ENHANCEMENT_STAGES[i];

    // Direct enhancement (critique is now embedded in the prompt)
    yield { type: 'enhancement', status: 'running', stageName: stage.name, stageIndex: i };
    currentText = await runEnhancementStage(i, topic, research, currentText);
    yield {
      type: 'enhancement',
      status: 'done',
      stageName: stage.name,
      stageIndex: i,
      text: currentText,
      enhancedText: currentText,
    };
  }

  // Pronunciation extraction (runs in parallel with nothing — just extract metadata)
  yield { type: 'pronunciation', status: 'running' };
  const pronunciations = await runPronunciation(currentText);
  yield { type: 'pronunciation', status: 'done', pronunciations };

  yield { type: 'done', finalText: currentText, pronunciations };
}

/**
 * Quick pipeline - faster generation with fewer refinements
 * Skips enhancement stages and support check for speed
 */
export async function* runQuickPipeline(
  topic: string,
  length: EpisodeLength = '10 min',
  style?: string
): AsyncGenerator<PipelineStep> {
  // Stage 0: Research (still essential for quality)
  yield { type: 'research', status: 'running' };
  const research = await runResearch(topic, length);
  yield { type: 'research', status: 'done', research };

  // Stage 1: Parallel Drafts (with style applied)
  yield { type: 'drafts', status: 'running' };
  const drafts = await runParallelDrafts(topic, research, length, style);
  yield { type: 'drafts', status: 'done', drafts, draftA: drafts[0].text, draftB: drafts[1].text };

  // Judge - pick the best draft
  yield { type: 'judge', status: 'running' };
  const judgeResult = await runJudge(topic, drafts);
  yield {
    type: 'judge',
    status: 'done',
    winner: judgeResult.winnerLabel,
    winnerText: judgeResult.winnerText,
    judgment: judgeResult.judgment,
  };

  // Skip enhancement stages - use winning draft directly
  // The winning draft is already good quality from Claude or GPT-4
  yield { type: 'done', finalText: judgeResult.winnerText };
}

// ============================================================================
// Learning Add-ons
// ============================================================================

export async function generateAddon(
  addonKey: keyof typeof LEARNING_ADDONS,
  topic: string,
  transcript: string
): Promise<string> {
  const addon = LEARNING_ADDONS[addonKey];
  return callLLM({
    provider: 'anthropic',
    system: addon.system,
    user: addon.userTemplate.replace('{topic}', topic).replace('{transcript}', transcript),
    temperature: addon.temperature,
    promptVersion: PROMPT_VERSION,
    purpose: `addon:${addonKey}`,
  });
}

// ============================================================================
// Canon Protocol — Remaster Pipeline
// ============================================================================

/**
 * Run the canon remaster: research → remaster draft → enhancement → quality gate.
 * Uses the existing transcript as a seed but rewrites to higher standards.
 */
export async function runCanonRemaster(
  topic: string,
  seedTranscript: string,
  length: EpisodeLength = '10 min'
): Promise<{
  finalText: string;
  research: string;
  qualityGate: { pass: boolean; scores: Record<string, number>; average: number; weakest?: string; suggestion?: string };
}> {
  // 1. Fresh research (canon episodes need the latest data)
  const research = await runResearch(topic, length);

  // 2. Canon remaster draft (uses seed transcript + fresh research)
  const remasterPrompt = getCanonRemasterPrompt(topic, research, seedTranscript, length);
  let remasteredText = await callLLM({
    provider: 'anthropic',
    system: remasterPrompt.system,
    user: remasterPrompt.user,
    temperature: remasterPrompt.temperature,
    promptVersion: PROMPT_VERSION,
    purpose: 'canon-remaster',
  });

  // 3. Support check
  const supportResult = await runSupportCheck(research, remasteredText);
  if (!supportResult.ok) {
    console.log('[Canon] Support check flagged issues, proceeding with enhancement to address');
  }

  // 4. Enhancement passes (same as regular pipeline)
  for (let i = 0; i < ENHANCEMENT_STAGES.length; i++) {
    remasteredText = await runEnhancementStage(i, topic, research, remasteredText);
  }

  // 5. Quality gate
  const qualityGate = await runCanonQualityGate(topic, remasteredText);

  return {
    finalText: remasteredText,
    research,
    qualityGate,
  };
}

/**
 * Run the canon quality gate evaluation on a script.
 */
export async function runCanonQualityGate(
  topic: string,
  script: string
): Promise<{
  pass: boolean;
  scores: Record<string, number>;
  average: number;
  weakest?: string;
  suggestion?: string;
}> {
  const result = await callLLM({
    provider: 'openai',
    model: FAST_OPENAI_MODEL,
    system: CANON_QUALITY_GATE.system,
    user: CANON_QUALITY_GATE.userTemplate
      .replace('{topic}', topic)
      .replace('{script}', script),
    temperature: CANON_QUALITY_GATE.temperature,
    maxTokens: 1024,
    promptVersion: PROMPT_VERSION,
    purpose: 'canon-quality-gate',
  });

  try {
    // Extract JSON from response (may have markdown fences)
    const jsonMatch = result.match(/\{[\s\S]*\}/);
    if (!jsonMatch) {
      console.warn('[Canon QG] Could not parse quality gate response');
      return { pass: false, scores: {}, average: 0, suggestion: 'Failed to parse quality gate' };
    }

    const parsed = JSON.parse(jsonMatch[0]);
    const scores: Record<string, number> = parsed.scores || {};
    const scoreValues = Object.values(scores).filter((v): v is number => typeof v === 'number');
    const average = scoreValues.length > 0
      ? scoreValues.reduce((a, b) => a + b, 0) / scoreValues.length
      : 0;

    const minScore = scoreValues.length > 0 ? Math.min(...scoreValues) : 0;
    const pass = average >= CANON_QUALITY_GATE.passThreshold &&
                 minScore >= CANON_QUALITY_GATE.minDimensionScore;

    return {
      pass,
      scores,
      average,
      weakest: parsed.weakest,
      suggestion: parsed.suggestion,
    };
  } catch (e) {
    console.warn('[Canon QG] JSON parse error:', e);
    return { pass: false, scores: {}, average: 0, suggestion: 'Quality gate parse error' };
  }
}
